fendermint/vm/interpreter/src/fvm/query.rs

```diff
@@ -99,3 +96,131 @@ where
         Ok((state, res))
     }
 }
+
+impl<DB> FvmMessageInterpreter<DB>
+where
+    DB: Blockstore + 'static + Send + Sync + Clone,
+{
+    /// Overestimation rate applied to gas to ensure that the
+    /// message goes through in the gas estimation.
+    const GAS_OVERESTIMATION_RATE: f64 = 1.25;
+    /// Default gas premium value. Inferred through a quick search through
+    /// InvokeEVM messages in filfox. The default value is only used if
+    /// the user hasn't specified a gas premium.
+    const DEFAULT_GAS_PREMIUM: u64 = 20000;
+    /// Gas search step increase used to find the optimal gas limit.
+    const GAS_SEARCH_STEP: f64 = 1.2;
+
+    fn estimate_gassed_msg(
+        &self,
+        state: &FvmQueryState<DB>,
+        msg: Box<Message>,
+    ) -> anyhow::Result<Message> {
+        let mut out = (*msg).clone();
+        // estimate the gas limit and assign it to the message
+        // do not reuse the cache
+        let ret = state.call_with_cache(*msg.clone(), false)?;
+        if ret.msg_receipt.exit_code != ExitCode::OK {
+            return Err(anyhow::anyhow!(
+                "message execution failed with error code {}",
+                ret.msg_receipt.exit_code
+            ));
+        }
+        out.gas_limit = (ret.msg_receipt.gas_used as f64 * Self::GAS_OVERESTIMATION_RATE) as u64;
+        if out.gas_premium.is_zero() {
+            // TODO: Instead of assigning a default value here, we should analyze historical
+            // blocks from the current height to estimate an accurate value for this premium.
+            // To achieve this we would need to perform a set of ABCI queries.
+            // In the meantime, this value should be good enough to make sure that the
+            // message is included in a block.
+            out.gas_premium = TokenAmount::from_nano(BigInt::from(Self::DEFAULT_GAS_PREMIUM));
+        }
+        if out.gas_fee_cap.is_zero() {
+            // Compute the fee cap from gas premium and applying an additional overestimation.
+            let overestimated_limit = (out.gas_limit as f64 * Self::GAS_OVERESTIMATION_RATE) as u64;
```

Comment:
This is now doubly over estimated. Is that intentional?

Reply:
Hm, I might be wrong, but we have this on line 129:

```rust
        out.gas_limit = (ret.msg_receipt.gas_used as f64 * Self::GAS_OVERESTIMATION_RATE) as u64;
```
and then on line 140:
```rust
let overestimated_limit = (out.gas_limit as f64 * Self::GAS_OVERESTIMATION_RATE) as u64;
```

Which gives:
```
 out.gas_fee_cap = out.gas_limit * GAS_OVERESTIMATION_RATE * GAS_OVERESTIMATION_RATE  + &out.gas_premium,
```

That's `GAS_OVERESTIMATION_RATE` squared.

Reply:
This was supposed to give additional fee cap space in case the estimation was too conservative. We are not setting a fee cap anymore so this comment doesn't apply anymore.